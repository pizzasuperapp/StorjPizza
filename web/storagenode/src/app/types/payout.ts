// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import {
    EstimatedPayout,
    PayoutPeriod,
    SatelliteHeldHistory,
    SatellitePayoutForPeriod,
    TotalPayments,
    TotalPaystubForPeriod,
} from '@/storagenode/payouts/payouts';

/**
 * Holds 'start' and 'end' of payout period range.
 */
export class PayoutInfoRange {
    public constructor(
        public start: PayoutPeriod | null = null,
        public end: PayoutPeriod = new PayoutPeriod(),
    ) {}
}

/**
 * Holds all payout module state.
 */
export class PayoutState {
    public constructor (
        public totalPaystubForPeriod: TotalPaystubForPeriod = new TotalPaystubForPeriod(),
        public periodRange: PayoutInfoRange = new PayoutInfoRange(),
        public totalPayments: TotalPayments = new TotalPayments(),
        public currentMonthEarnings: number = 0,
        public heldPercentage: number = 0,
        public payoutPeriods: PayoutPeriod[] = [],
        public heldHistory: SatelliteHeldHistory[] = [],
        public payoutHistory: SatellitePayoutForPeriod[] = [],
        public payoutHistoryPeriod: string = '',
        public estimation: EstimatedPayout = new EstimatedPayout(),
        public payoutHistoryAvailablePeriods: PayoutPeriod[] = [],
    ) {}
}

export interface StoredMonthsByYear {
    [key: number]: MonthButton[];
}

/**
 * Holds all months names.
 */
export const monthNames = [
    'January', 'February', 'March', 'April',
    'May', 'June', 'July',	'August',
    'September', 'October', 'November',	'December',
];

/**
 * Describes month button entity for calendar.
 */
export class MonthButton {
    public constructor(
        public year: number = 0,
        public index: number = 0,
        public active: boolean = false,
        public selected: boolean = false,
    ) {}

    /**
     * Returns month label depends on index.
     */
    public get name(): string {
        return monthNames[this.index] ? monthNames[this.index].slice(0, 3) : '';
    }
}
