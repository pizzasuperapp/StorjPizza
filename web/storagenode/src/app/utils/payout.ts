// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

// TODO: functions should be moved to related business logic layer
/**
 * Returns held percentage depends on number of months that node is online.
 * @param startedAt date since node is online.
 */
export function getHeldPercentage(startedAt: Date): number {
    const monthsOnline = getMonthsBeforeNow(startedAt);

    switch (true) {
    case monthsOnline < 4:
        return 75;
    case monthsOnline < 7:
        return 50;
    case monthsOnline < 10:
        return 25;
    default:
        return 0;
    }
}

/**
 * Returns number of months passes till now.
 * @param startedAt date since node is online.
 */
export function getMonthsBeforeNow(startedAt: Date): number {
    const now = new Date();
    const yearsDiff =  now.getUTCFullYear() - startedAt.getUTCFullYear();

    return (yearsDiff * 12) + (now.getUTCMonth() - startedAt.getUTCMonth()) + 1;
}
