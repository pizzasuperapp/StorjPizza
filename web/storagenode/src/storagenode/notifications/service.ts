// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import { NotificationsApi, NotificationsCursor, NotificationsResponse } from '@/storagenode/notifications/notifications';

/**
 * PayoutService is used to store and handle node paystub information.
 * PayoutService exposes a business logic related to payouts.
 */
export class NotificationsService {
    private readonly api: NotificationsApi;

    public constructor(api: NotificationsApi) {
        this.api = api;
    }

    /**
     * Fetch notifications.
     *
     * @returns notifications response.
     * @throws Error
     */
    public async notifications(index: number, limit?: number): Promise<NotificationsResponse> {
        const cursor = new NotificationsCursor(index, limit);

        return await this.api.get(cursor);
    }

    /**
     * Marks single notification as read on server.
     * @param id
     */
    public async readSingeNotification(id: string): Promise<void> {
        await this.api.read(id);
    }

    /**
     * Marks all notifications as read on server.
     */
    public async readAllNotifications(): Promise<void> {
        await this.api.readAll();
    }
}
