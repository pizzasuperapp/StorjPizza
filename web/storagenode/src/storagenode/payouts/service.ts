// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import {
    EstimatedPayout,
    PaymentInfoParameters,
    PayoutApi,
    PayoutPeriod,
    Paystub,
    SatelliteHeldHistory,
    SatellitePayoutForPeriod,
    TotalPayments,
    TotalPaystubForPeriod,
} from '@/storagenode/payouts/payouts';

/**
 * PayoutService is used to store and handle node paystub information.
 * PayoutService exposes a business logic related to payouts.
 */
export class PayoutService {
    private readonly payouts: PayoutApi;

    public constructor(api: PayoutApi) {
        this.payouts = api;
    }

    /**
     * Gets summary of paystubs for given period.
     * @param start period start
     * @param end period end
     * @param satelliteId
     */
    public async paystubSummaryForPeriod(start: PayoutPeriod | null, end: PayoutPeriod, satelliteId?: string): Promise<TotalPaystubForPeriod> {
        const paystubs: Paystub[] = await this.payouts.getPaystubsForPeriod(new PaymentInfoParameters(start, end, satelliteId));

        return new TotalPaystubForPeriod(paystubs);
    }

    /**
     * Gets held and paid summary for given period.
     * @param start period start
     * @param end period end
     * @param satelliteId
     */
    public async totalPayments(start: PayoutPeriod, end: PayoutPeriod, satelliteId: string): Promise<TotalPayments> {
        const paystubs: Paystub[] = await this.payouts.getPaystubsForPeriod(new PaymentInfoParameters(start, end, satelliteId));

        return new TotalPayments(paystubs);
    }

    /**
     * Gets list of payout periods that have paystubs for selected satellite.
     * If satelliteId is not provided returns periods for all satellites.
     * @param satelliteId
     */
    public async availablePeriods(satelliteId: string): Promise<PayoutPeriod[]> {
        return await this.payouts.getPayoutPeriods(satelliteId);
    }

    /**
     * Gets list of payout history items for given period by satellites.
     * @param payoutHistoryPeriod year and month representation
     */
    public async payoutHistory(payoutHistoryPeriod: string): Promise<SatellitePayoutForPeriod[]> {
        return await this.payouts.getPayoutHistory(payoutHistoryPeriod);
    }

    /**
     * Gets list of held history for all satellites.
     */
    public async allSatellitesHeldHistory(): Promise<SatelliteHeldHistory[]> {
        return await this.payouts.getHeldHistory();
    }

    /**
     * Gets estimated payout when no data in paystub.
     * @param satelliteId
     */
    public async estimatedPayout(satelliteId: string): Promise<EstimatedPayout> {
        return await this.payouts.getEstimatedPayout(satelliteId);
    }
}
