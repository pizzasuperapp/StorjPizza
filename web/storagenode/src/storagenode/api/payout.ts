// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import {
    EstimatedPayout,
    PaymentInfoParameters,
    PayoutApi,
    PayoutPeriod,
    Paystub,
    PreviousMonthEstimatedPayout,
    SatelliteHeldHistory,
    SatellitePayoutForPeriod,
} from '@/storagenode/payouts/payouts';
import { HttpClient } from '@/storagenode/utils/httpClient';

/**
 * PayoutHttpApi is a http implementation of Payout API.
 * Exposes all payout-related functionality
 */
export class PayoutHttpApi implements PayoutApi {
    private readonly client: HttpClient = new HttpClient();
    private readonly ROOT_PATH: string = '/api/heldamount';

    /**
     * Fetch paystubs for selected period.
     *
     * @returns paystubs for given period
     * @throws Error
     */
    public async getPaystubsForPeriod(paymentInfoParameters: PaymentInfoParameters): Promise<Paystub[]> {
        let path = `${this.ROOT_PATH}/paystubs/`;

        if (paymentInfoParameters.start) {
            path += paymentInfoParameters.start.period + '/';
        }

        path += paymentInfoParameters.end.period;

        if (paymentInfoParameters.satelliteId) {
            path += '?id=' + paymentInfoParameters.satelliteId;
        }

        const response = await this.client.get(path);

        if (!response.ok) {
            throw new Error('can not get held information');
        }

        const responseBody = await response.json() || [];
        const data: any[] = !Array.isArray(responseBody) ? [ responseBody ] : responseBody; // eslint-disable-line @typescript-eslint/no-explicit-any

        return data.map((paystubJson: any) => { // eslint-disable-line @typescript-eslint/no-explicit-any
            return new Paystub(
                paystubJson.usageAtRest,
                paystubJson.usageGet,
                paystubJson.usagePut,
                paystubJson.usageGetRepair,
                paystubJson.usagePutRepair,
                paystubJson.usageGetAudit,
                paystubJson.compAtRest,
                paystubJson.compGet,
                paystubJson.compPut,
                paystubJson.compGetRepair,
                paystubJson.compPutRepair,
                paystubJson.compGetAudit,
                paystubJson.surgePercent,
                paystubJson.held,
                paystubJson.owed,
                paystubJson.disposed,
                paystubJson.paid,
                paystubJson.distributed,
            );
        });
    }

    /**
     * Fetches available payout periods.
     *
     * @returns payout periods list
     * @throws Error
     */
    public async getPayoutPeriods(id: string): Promise<PayoutPeriod[]> {
        let path = `${this.ROOT_PATH}/periods`;

        if (id) {
            path += '?id=' + id;
        }

        const response = await this.client.get(path);

        if (!response.ok) {
            throw new Error('can not get payout periods');
        }

        const result = await response.json() || [];

        return result.map(period => {
            return PayoutPeriod.fromString(period);
        });
    }

    /**
     * Fetch payout history for given period.
     *
     * @returns payout information
     * @throws Error
     */
    public async getPayoutHistory(period: string): Promise<SatellitePayoutForPeriod[]> {
        const path = `${this.ROOT_PATH}/payout-history/${period}`;

        const response = await this.client.get(path);

        if (!response.ok) {
            throw new Error('can not get payout history information');
        }

        const data: any = await response.json() || []; // eslint-disable-line @typescript-eslint/no-explicit-any

        return data.map((payoutHistoryItem: any) => { // eslint-disable-line @typescript-eslint/no-explicit-any
            return new SatellitePayoutForPeriod(
                payoutHistoryItem.satelliteID,
                payoutHistoryItem.satelliteURL,
                payoutHistoryItem.age,
                payoutHistoryItem.earned,
                payoutHistoryItem.surge,
                payoutHistoryItem.surgePercent,
                payoutHistoryItem.held,
                payoutHistoryItem.afterHeld,
                payoutHistoryItem.disposed,
                payoutHistoryItem.paid,
                payoutHistoryItem.receipt,
                payoutHistoryItem.isExitComplete,
                payoutHistoryItem.heldPercent,
                payoutHistoryItem.distributed,
            );
        });
    }

    /**
     * Fetch total payout information.
     *
     * @returns total payout information
     * @throws Error
     */
    public async getHeldHistory(): Promise<SatelliteHeldHistory[]> {
        const path = `${this.ROOT_PATH}/held-history`;

        const response = await this.client.get(path);

        if (!response.ok) {
            throw new Error('can not get held history information');
        }

        const data: any = await response.json() || []; // eslint-disable-line @typescript-eslint/no-explicit-any

        return data.map((historyItem: any) => { // eslint-disable-line @typescript-eslint/no-explicit-any
            return new SatelliteHeldHistory(
                historyItem.satelliteID,
                historyItem.satelliteName,
                historyItem.holdForFirstPeriod,
                historyItem.holdForSecondPeriod,
                historyItem.holdForThirdPeriod,
                historyItem.totalHeld,
                historyItem.totalDisposed,
                new Date(historyItem.joinedAt),
            );
        });
    }

    /**
     * Fetch estimated payout information.
     *
     * @returns estimated payout information
     * @throws Error
     */
    public async getEstimatedPayout(satelliteId: string): Promise<EstimatedPayout> {
        let path = '/api/sno/estimated-payout';

        if (satelliteId) {
            path += '?id=' + satelliteId;
        }

        const response = await this.client.get(path);

        if (!response.ok) {
            throw new Error('can not get estimated payout information');
        }

        const data: any = await response.json() || new EstimatedPayout(); // eslint-disable-line @typescript-eslint/no-explicit-any

        return new EstimatedPayout(
            new PreviousMonthEstimatedPayout(
                data.currentMonth.egressBandwidth,
                data.currentMonth.egressBandwidthPayout,
                data.currentMonth.egressRepairAudit,
                data.currentMonth.egressRepairAuditPayout,
                data.currentMonth.diskSpace,
                data.currentMonth.diskSpacePayout,
                data.currentMonth.heldRate,
                data.currentMonth.payout,
                data.currentMonth.held,
            ),
            new PreviousMonthEstimatedPayout(
                data.previousMonth.egressBandwidth,
                data.previousMonth.egressBandwidthPayout,
                data.previousMonth.egressRepairAudit,
                data.previousMonth.egressRepairAuditPayout,
                data.previousMonth.diskSpace,
                data.previousMonth.diskSpacePayout,
                data.previousMonth.heldRate,
                data.previousMonth.payout,
                data.previousMonth.held,
            ),
            data.currentMonthExpectations,
        );
    }
}
