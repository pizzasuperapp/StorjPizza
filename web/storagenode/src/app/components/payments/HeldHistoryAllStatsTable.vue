// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <div>
        <div class="held-history-table-container--large">
            <div class="held-history-table-container--large__labels-area">
                <div class="column justify-start column-1">
                    <p class="held-history-table-container--large__labels-area__text">Satellite</p>
                </div>
                <div class="column justify-end column-2">
                    <p class="held-history-table-container--large__labels-area__text">First Contact</p>
                </div>
                <div class="column justify-end column-3">
                    <p class="held-history-table-container--large__labels-area__text">Held Total</p>
                </div>
                <div class="column justify-end column-4">
                    <p class="held-history-table-container--large__labels-area__text">Held Returned</p>
                </div>
            </div>
            <div v-for="item in allSatellitesHeldHistory" :key="item.satelliteID" class="held-history-table-container--large__info-area">
                <div class="justify-start column-1">
                    <p class="held-history-table-container--large__info-area__text">{{ item.satelliteName }}</p>
                    <p class="held-history-table-container--large__info-area__months">{{ item.monthsWithNode }} month</p>
                </div>
                <div class="column justify-end column-2">
                    <p class="held-history-table-container--large__info-area__text">{{ item.joinedAt.toISOString().split('T')[0] }}</p>
                </div>
                <div class="column justify-end column-3">
                    <p class="held-history-table-container--large__info-area__text">{{ item.totalHeld | centsToDollars }}</p>
                </div>
                <div class="column justify-end column-4">
                    <p class="held-history-table-container--large__info-area__text">{{ item.totalDisposed | centsToDollars }}</p>
                </div>
            </div>
        </div>
        <div class="held-history-table-container--small">
            <HeldHistoryAllStatsTableItemSmall
                v-for="item in allSatellitesHeldHistory"
                :key="item.satelliteID"
                :held-history-item="item"
            />
        </div>
    </div>
</template>

<script lang="ts">
import { Component } from 'vue-property-decorator';

import BaseHeldHistoryTable from '@/app/components/payments/BaseHeldHistoryTable.vue';
import HeldHistoryAllStatsTableItemSmall from '@/app/components/payments/HeldHistoryAllStatsTableItemSmall.vue';

import { SatelliteHeldHistory } from '@/storagenode/payouts/payouts';

// @vue/component
@Component({
    components: {
        HeldHistoryAllStatsTableItemSmall,
    },
})
export default class HeldHistoryAllStatsTable extends BaseHeldHistoryTable {
    /**
     * Returns list of satellite held history items by periods from store.
     */
    public get allSatellitesHeldHistory(): SatelliteHeldHistory[] {
        return this.$store.state.payoutModule.heldHistory;
    }
}
</script>
