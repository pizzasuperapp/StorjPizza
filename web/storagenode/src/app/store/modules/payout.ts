// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import { StoreModule } from '@/app/store';
import {
    PayoutInfoRange,
    PayoutState,
} from '@/app/types/payout';
import { StorageNodeState } from '@/app/types/sno';
import { getHeldPercentage } from '@/app/utils/payout';
import { SizeBreakpoints } from '@/private/memory/size';
import {
    EstimatedPayout,
    PayoutPeriod,
    SatelliteHeldHistory,
    SatellitePayoutForPeriod,
    TotalPayments,
    TotalPaystubForPeriod,
} from '@/storagenode/payouts/payouts';
import { PayoutService } from '@/storagenode/payouts/service';

export const PAYOUT_MUTATIONS = {
    SET_PAYOUT_INFO: 'SET_PAYOUT_INFO',
    SET_RANGE: 'SET_RANGE',
    SET_TOTAL: 'SET_TOTAL',
    SET_HELD_PERCENT: 'SET_HELD_PERCENT',
    SET_HELD_HISTORY: 'SET_HELD_HISTORY',
    SET_ESTIMATION: 'SET_ESTIMATION',
    SET_PERIODS: 'SET_PERIODS',
    SET_PAYOUT_HISTORY: 'SET_PAYOUT_HISTORY',
    SET_PAYOUT_HISTORY_PERIOD: 'SET_PAYOUT_HISTORY_PERIOD',
    SET_PAYOUT_HISTORY_AVAILABLE_PERIODS: 'SET_PAYOUT_HISTORY_AVAILABLE_PERIODS',
};

export const PAYOUT_ACTIONS = {
    GET_PAYOUT_INFO: 'GET_PAYOUT_INFO',
    SET_PERIODS_RANGE: 'SET_PERIODS_RANGE',
    GET_TOTAL: 'GET_TOTAL',
    GET_HELD_HISTORY: 'GET_HELD_HISTORY',
    GET_ESTIMATION: 'GET_ESTIMATION',
    GET_PERIODS: 'GET_PERIODS',
    GET_PAYOUT_HISTORY: 'GET_PAYOUT_HISTORY',
    SET_PAYOUT_HISTORY_PERIOD: 'SET_PAYOUT_HISTORY_PERIOD',
};

// TODO: move to config in storagenode/payouts
export const BANDWIDTH_DOWNLOAD_PRICE_PER_TB = 2000;
export const BANDWIDTH_REPAIR_PRICE_PER_TB = 1000;
export const DISK_SPACE_PRICE_PER_TB = 150;

interface PayoutContext {
    rootState: {
        node: StorageNodeState;
    };
    state: PayoutState;
    commit: (string, ...unknown) => void;
}

/**
 * creates notifications module with all dependencies
 *
 * @param service - payments service
 */
export function newPayoutModule(service: PayoutService): StoreModule<PayoutState> {
    return {
        state: new PayoutState(),
        mutations: {
            [PAYOUT_MUTATIONS.SET_PAYOUT_INFO](state: PayoutState, totalPaystubForPeriod: TotalPaystubForPeriod): void {
                state.totalPaystubForPeriod = totalPaystubForPeriod;
            },
            [PAYOUT_MUTATIONS.SET_TOTAL](state: PayoutState, totalPayments: TotalPayments): void {
                state.totalPayments = totalPayments;
                state.currentMonthEarnings = totalPayments.currentMonthEarnings;
            },
            [PAYOUT_MUTATIONS.SET_RANGE](state: PayoutState, periodRange: PayoutInfoRange): void {
                state.periodRange = periodRange;
            },
            [PAYOUT_MUTATIONS.SET_HELD_PERCENT](state: PayoutState, heldPercentage: number): void {
                state.heldPercentage = heldPercentage;
            },
            [PAYOUT_MUTATIONS.SET_HELD_HISTORY](state: PayoutState, heldHistory: SatelliteHeldHistory[]): void {
                state.heldHistory = heldHistory;
            },
            [PAYOUT_MUTATIONS.SET_ESTIMATION](state: PayoutState, estimatedInfo: EstimatedPayout): void {
                state.estimation = estimatedInfo;
            },
            [PAYOUT_MUTATIONS.SET_PERIODS](state: PayoutState, periods: PayoutPeriod[]): void {
                state.payoutPeriods = periods;
            },
            [PAYOUT_MUTATIONS.SET_PAYOUT_HISTORY](state: PayoutState, payoutHistory: SatellitePayoutForPeriod[]): void {
                state.payoutHistory = payoutHistory;
            },
            [PAYOUT_MUTATIONS.SET_PAYOUT_HISTORY_PERIOD](state: PayoutState, period: string): void {
                state.payoutHistoryPeriod = period;
            },
            [PAYOUT_MUTATIONS.SET_PAYOUT_HISTORY_AVAILABLE_PERIODS](state: PayoutState, periods: PayoutPeriod[]): void {
                state.payoutHistoryAvailablePeriods = periods;
            },
        },
        actions: {
            [PAYOUT_ACTIONS.GET_PAYOUT_INFO]: async function ({ commit, state, rootState }: PayoutContext, satelliteId = ''): Promise<void> {
                const totalPaystubForPeriod = await service.paystubSummaryForPeriod(
                    state.periodRange.start,
                    state.periodRange.end,
                    satelliteId,
                );

                commit(PAYOUT_MUTATIONS.SET_HELD_PERCENT, getHeldPercentage(rootState.node.selectedSatellite.joinDate));
                commit(PAYOUT_MUTATIONS.SET_PAYOUT_INFO, totalPaystubForPeriod);
            },
            [PAYOUT_ACTIONS.GET_TOTAL]: async function ({ commit, rootState }: PayoutContext, satelliteId = ''): Promise<void> {
                const now = new Date();
                const start = new PayoutPeriod(rootState.node.selectedSatellite.joinDate.getUTCFullYear(), rootState.node.selectedSatellite.joinDate.getUTCMonth());
                const end = new PayoutPeriod(now.getUTCFullYear(), now.getUTCMonth());

                const totalPayments = await service.totalPayments(start, end, satelliteId);

                // TODO: move to service
                const currentBandwidthDownload = (rootState.node.egressChartData || [])
                    .map(data => data.egress.usage)
                    .reduce((previous, current) => previous + current, 0);

                const currentBandwidthAuditAndRepair = (rootState.node.egressChartData || [])
                    .map(data => data.egress.audit + data.egress.repair)
                    .reduce((previous, current) => previous + current, 0);

                const approxHourInMonth = 720;
                const currentDiskSpace = (rootState.node.storageChartData || [])
                    .map(data => data.atRestTotal)
                    .reduce((previous, current) => previous + current, 0) / approxHourInMonth;

                const thisMonthEarnings = (currentBandwidthDownload * BANDWIDTH_DOWNLOAD_PRICE_PER_TB
                    + currentBandwidthAuditAndRepair * BANDWIDTH_REPAIR_PRICE_PER_TB
                    + currentDiskSpace * DISK_SPACE_PRICE_PER_TB) / SizeBreakpoints.TB;

                totalPayments.setCurrentMonthEarnings(thisMonthEarnings);

                commit(PAYOUT_MUTATIONS.SET_HELD_PERCENT, getHeldPercentage(rootState.node.selectedSatellite.joinDate));
                commit(PAYOUT_MUTATIONS.SET_TOTAL, totalPayments);
            },
            [PAYOUT_ACTIONS.SET_PERIODS_RANGE]: function ({ commit }: PayoutContext, periodRange: PayoutInfoRange): void {
                commit(PAYOUT_MUTATIONS.SET_RANGE, periodRange);
            },
            [PAYOUT_ACTIONS.GET_HELD_HISTORY]: async function ({ commit }: PayoutContext): Promise<void> {
                const heldHistory = await service.allSatellitesHeldHistory();

                commit(PAYOUT_MUTATIONS.SET_HELD_HISTORY, heldHistory);
            },
            [PAYOUT_ACTIONS.GET_PERIODS]: async function ({commit}: PayoutContext, satelliteId = ''): Promise<void> {
                const periods = await service.availablePeriods(satelliteId);

                commit(PAYOUT_MUTATIONS.SET_PERIODS, periods);

                if (!satelliteId) {
                    commit(PAYOUT_MUTATIONS.SET_PAYOUT_HISTORY_AVAILABLE_PERIODS, periods);
                }
            },
            [PAYOUT_ACTIONS.GET_ESTIMATION]: async function ({ commit }: PayoutContext, satelliteId = ''): Promise<void> {
                const estimatedInfo = await service.estimatedPayout(satelliteId);

                commit(PAYOUT_MUTATIONS.SET_ESTIMATION, estimatedInfo);
            },
            [PAYOUT_ACTIONS.GET_PAYOUT_HISTORY]: async function ({ commit, state }: PayoutContext): Promise<void> {
                if (!state.payoutHistoryPeriod) return;

                const payoutHistory = await service.payoutHistory(state.payoutHistoryPeriod);

                commit(PAYOUT_MUTATIONS.SET_PAYOUT_HISTORY, payoutHistory);
            },
            [PAYOUT_ACTIONS.SET_PAYOUT_HISTORY_PERIOD]: function ({ commit }: PayoutContext, period: string): void {
                commit(PAYOUT_MUTATIONS.SET_PAYOUT_HISTORY_PERIOD, period);
            },
        },
        getters: {
            totalPaidForPayoutHistoryPeriod: (state: PayoutState): number => {
                return state.payoutHistory.map(data => data.paid)
                    .reduce((previous, current) => previous + current, 0);
            },
        },
    };
}
