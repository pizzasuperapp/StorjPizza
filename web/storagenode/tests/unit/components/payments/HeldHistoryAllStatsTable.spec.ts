// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import Vuex from 'vuex';

import HeldHistoryAllStatsTable from '@/app/components/payments/HeldHistoryAllStatsTable.vue';

import { newPayoutModule, PAYOUT_MUTATIONS } from '@/app/store/modules/payout';
import { PayoutHttpApi } from '@/storagenode/api/payout';
import { SatelliteHeldHistory } from '@/storagenode/payouts/payouts';
import { PayoutService } from '@/storagenode/payouts/service';
import { createLocalVue, shallowMount } from '@vue/test-utils';

const localVue = createLocalVue();
localVue.use(Vuex);

localVue.filter('centsToDollars', (cents: number): string => {
    return `$${(cents / 100).toFixed(2)}`;
});

const payoutApi = new PayoutHttpApi();
const payoutService = new PayoutService(payoutApi);
const payoutModule = newPayoutModule(payoutService);

const store = new Vuex.Store({ modules: { payoutModule }});

describe('HeldHistoryAllStatsTable', (): void => {
    it('renders correctly with actual values', async (): Promise<void> => {
        const _Date = Date;
        const mockedDate = new Date(1580522290000);
        const testJoinAt = new Date(Date.UTC(2019, 6, 30));
        global.Date = jest.fn(() => mockedDate);

        const wrapper = shallowMount(HeldHistoryAllStatsTable, {
            store,
            localVue,
        });

        const testHeldHistory = [
            new SatelliteHeldHistory('1', 'name1', 1, 50000, 0, 1, 1, testJoinAt),
            new SatelliteHeldHistory('2', 'name2', 5, 50000, 422280, 0, 0, testJoinAt),
            new SatelliteHeldHistory('3', 'name3', 6, 50000, 7333880, 7852235, 0, testJoinAt),
        ];

        await store.commit(PAYOUT_MUTATIONS.SET_HELD_HISTORY, testHeldHistory);

        expect(wrapper).toMatchSnapshot();

        global.Date = _Date;
    });
});
