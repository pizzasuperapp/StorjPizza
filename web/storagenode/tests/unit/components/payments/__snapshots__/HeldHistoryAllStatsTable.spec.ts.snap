// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeldHistoryAllStatsTable renders correctly with actual values 1`] = `
<div>
  <div class="held-history-table-container--large">
    <div class="held-history-table-container--large__labels-area">
      <div class="column justify-start column-1">
        <p class="held-history-table-container--large__labels-area__text">Satellite</p>
      </div>
      <div class="column justify-end column-2">
        <p class="held-history-table-container--large__labels-area__text">First Contact</p>
      </div>
      <div class="column justify-end column-3">
        <p class="held-history-table-container--large__labels-area__text">Held Total</p>
      </div>
      <div class="column justify-end column-4">
        <p class="held-history-table-container--large__labels-area__text">Held Returned</p>
      </div>
    </div>
    <div class="held-history-table-container--large__info-area">
      <div class="justify-start column-1">
        <p class="held-history-table-container--large__info-area__text">name1</p>
        <p class="held-history-table-container--large__info-area__months">8 month</p>
      </div>
      <div class="column justify-end column-2">
        <p class="held-history-table-container--large__info-area__text">2019-07-30</p>
      </div>
      <div class="column justify-end column-3">
        <p class="held-history-table-container--large__info-area__text">$0.00</p>
      </div>
      <div class="column justify-end column-4">
        <p class="held-history-table-container--large__info-area__text">$0.00</p>
      </div>
    </div>
    <div class="held-history-table-container--large__info-area">
      <div class="justify-start column-1">
        <p class="held-history-table-container--large__info-area__text">name2</p>
        <p class="held-history-table-container--large__info-area__months">8 month</p>
      </div>
      <div class="column justify-end column-2">
        <p class="held-history-table-container--large__info-area__text">2019-07-30</p>
      </div>
      <div class="column justify-end column-3">
        <p class="held-history-table-container--large__info-area__text">$0.00</p>
      </div>
      <div class="column justify-end column-4">
        <p class="held-history-table-container--large__info-area__text">$0.00</p>
      </div>
    </div>
    <div class="held-history-table-container--large__info-area">
      <div class="justify-start column-1">
        <p class="held-history-table-container--large__info-area__text">name3</p>
        <p class="held-history-table-container--large__info-area__months">8 month</p>
      </div>
      <div class="column justify-end column-2">
        <p class="held-history-table-container--large__info-area__text">2019-07-30</p>
      </div>
      <div class="column justify-end column-3">
        <p class="held-history-table-container--large__info-area__text">$7.85</p>
      </div>
      <div class="column justify-end column-4">
        <p class="held-history-table-container--large__info-area__text">$0.00</p>
      </div>
    </div>
  </div>
  <div class="held-history-table-container--small">
    <heldhistoryallstatstableitemsmall-stub heldhistoryitem="[object Object]"></heldhistoryallstatstableitemsmall-stub>
    <heldhistoryallstatstableitemsmall-stub heldhistoryitem="[object Object]"></heldhistoryallstatstableitemsmall-stub>
    <heldhistoryallstatstableitemsmall-stub heldhistoryitem="[object Object]"></heldhistoryallstatstableitemsmall-stub>
  </div>
</div>
`;
